vault/resource_ad_secret_roles.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/hashicorp/vault/api"
+	"github.com/terraform-providers/terraform-provider-vault/util"
+)
+
+var (
+	adSecretBackendFromPathRegex         = regexp.MustCompile("^(.+)/roles/.+$")
+	adSecretBackendRoleNameFromPathRegex = regexp.MustCompile("^.+/roles/(.+$)")
+)
+
+func adSecretBackendRoleResource() *schema.Resource {
+	fields := map[string]*schema.Schema{
+		"backend": {
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "The mount path for the AD backend.",
+			StateFunc: func(v interface{}) string {
+				return strings.Trim(v.(string), "/")
+			},
+		},
+		"role": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `Name of the role.`,
+			ForceNew:    true,
+		},
+		"last_vault_rotation": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Last time Vault rotated this service account's password.",
+		},
+		"password_last_set": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Last time Vault set this service account's password.",
+		},
+		"service_account_name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `The username/logon name for the service account with which this role will be associated.`,
+		},
+		"ttl": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Description: `In seconds, the default password time-to-live.`,
+		},
+	}
+	return &schema.Resource{
+		Create: createRoleResource,
+		Update: updateRoleResource,
+		Read:   readRoleResource,
+		Delete: deleteRoleResource,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: fields,
+	}
+}
+
+func createRoleResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+	rolePath := fmt.Sprintf("%s/roles/%s", backend, role)
+
+	log.Printf("[DEBUG] Creating %q", rolePath)
+
+	data := map[string]interface{}{}
+	if v, ok := d.GetOkExists("service_account_name"); ok {
+		data["service_account_name"] = v
+	}
+
+	if v, ok := d.GetOkExists("ttl"); ok {
+		data["ttl"] = v
+	}
+
+	log.Printf("[DEBUG] Writing %q", rolePath)
+	if _, err := client.Logical().Write(rolePath, data); err != nil {
+		return fmt.Errorf("error writing %q: %s", rolePath, err)
+	}
+	d.SetId(rolePath)
+	log.Printf("[DEBUG] Wrote %q", rolePath)
+	return readRoleResource(d, meta)
+}
+
+func readRoleResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	rolePath := d.Id()
+	log.Printf("[DEBUG] Reading %q", rolePath)
+
+	roleName, err := adSecretBackendRoleNameFromPath(rolePath)
+	if err != nil {
+		log.Printf("[WARN] Removing AD role %q because its ID is invalid", rolePath)
+		d.SetId("")
+		return fmt.Errorf("invalid role ID %q: %s", rolePath, err)
+	}
+	d.Set("role", roleName)
+
+	backend, err := adSecretBackendFromPath(rolePath)
```

Comment:
How come `createRoleResource` and `updateRoleResource` get the backend string from `d.Get("backend")`, but read infers it from the role path?

Reply:
Ah, nevermind. This is the pattern on other role-related resources as well.